amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+    this_denominator_power = supported_denominators.index(denominator)
+    max_denominator_power = supported_denominators.index(minimum_pulse)
+
+    for this_denominator in supported_denominators[
+        this_denominator_power : max_denominator_power + 1
+    ]:
+        this_ql = 4 / this_denominator
+        starts = starts_from_lengths(measure_length, this_ql)
+        start_hierarchy.append(starts)
+
+    return start_hierarchy
+
+
+def starts_from_time_signature_and_levels(
+    ts_str: str,
+    levels: Union[list, None] = None,
+):
+    """
+    Gets starts from a time signature and a list of levels.
+    Records the starts associated with each level as a list,
+    and return a list of those lists.
+
+    Levels are defined by the hierarchies recorded in the time signature.
+    The 0th level for the full cycle (start 0.0 and full measure length) is always included.
+    "1" stands for the 1st level below that of the full metrical cycle
+    (e.g., a half cycle of 4/4 and similar time signtures like 2/2).
+    "2" is the next level down, (quarter cycle) and so on.
+
+    The function arranges a list of levels in increasing order:
+    level 0 is always included, and
+    the maximum permitted level (depth) is 6.
+
+    Level choices do not need to be successive:
+    e.g., in 4/4 a user could choose [1, 3],
+    with 1 standing for the half note level,
+    3 for the eight note level,
+    and skipping in intervening level 2 (quarter note).
+
+    Parameters
+    ----------
+    ts_str
+        Any valid string repersenting a time signature. See examples below.
+    levels
+        Optionally specify which levels implied by the time signature to use.
+
+    Returns
+    -------
+    list
+        Returns a list of lists with start positions by level.
+
+    Examples
+    --------
+
+    >>> starts_from_time_signature_and_levels("6/8", [1, 2])
+    [[0.0, 3.0], [0.0, 1.5, 3.0], [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0]]
+
+    >>> level_not_specified = starts_from_time_signature_and_levels("6/8")
+    >>> len(level_not_specified)
+    6
+
+    >>> level_not_specified[3]
+    [0.0, 0.25, 0.5, 0.75, 1.0, 1.25, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0]
+
+    """
+    if levels is None:
+        levels = range(6)
+
+    if max(levels) > 6:
+        raise ValueError("6 is the maximum level depth supported.")
+    levels = sorted(levels)  # smallest first
+    if 0 not in levels:  # likely, and not a problem
+        levels = [0] + levels  # Always have level 0 as first entry
+
+    starts_by_level = []
+
+    full_hierarchy = starts_from_time_signature(ts_str)
+    for level in levels:
+        starts_by_level.append(full_hierarchy[level])
+
+    return starts_by_level
+
+
+def starts_from_pulse_lengths(
+    pulse_lengths: list,
+    measure_length: Optional[float] = None,
+    require_2_or_3_between_levels: bool = False,
+):
+    """
+    Convert a list of pulse lengths into a corresponding list of lists
```

Comment:
Clarify that the user is providing the pulse lengths for each level of the metrical hierarchy, and the algorithm is expanding this into a hierarchy assuming isochrony.